websocket_server/BackupManager.js

```diff
@@ -0,0 +1,329 @@
+	 * @throws {Error} If lock cannot be acquired within timeout period
+	 */
+	async acquireLock(roomId) {
+		const startTime = Date.now()
+		while (this.locks.get(roomId)) {
+			if (Date.now() - startTime > this.lockTimeout) {
+				throw new Error(`Lock acquisition timeout for room ${roomId}`)
+			}
+			await new Promise((resolve) =>
+				setTimeout(resolve, this.lockRetryInterval),
+			)
+		}
+		this.locks.set(roomId, Date.now())
+	}
+
+	/**
+	 * Releases a lock for a specific room
+	 * @param {number} roomId - The room ID to unlock
+	 */
+	async releaseLock(roomId) {
+		this.locks.delete(roomId)
+	}
+
+	/**
+	 * Ensures roomId is a valid number
+	 * @param {number|string} roomId - The room ID to validate
+	 * @return {number} The validated room ID
+	 * @throws {Error} If roomId is invalid
+	 */
+	sanitizeRoomId(roomId) {
+		const numericRoomId = Number(roomId)
+		if (isNaN(numericRoomId) || numericRoomId <= 0) {
+			throw new Error('Invalid room ID: must be a positive number')
+		}
+		return numericRoomId
+	}
+
+	/**
+	 * Calculates SHA-256 checksum of data
+	 * @param {string | object} data - Data to calculate checksum for
+	 * @return {string} Hex string of SHA-256 hash
+	 */
+	calculateChecksum(data) {
+		return crypto
+			.createHash('sha256')
+			.update(typeof data === 'string' ? data : JSON.stringify(data))
+			.digest('hex')
+	}
+
+	/**
+	 * Creates a new backup for a room
+	 * @param {number} roomId - The room ID
+	 * @param {object} data - The data to backup
+	 * @return {Promise<string>} The backup ID
+	 * @throws {Error} If backup creation fails
+	 */
+	async createBackup(roomId, data) {
+		if (!roomId || !data) {
+			throw new Error('Invalid backup parameters')
+		}
+
+		const sanitizedRoomId = this.sanitizeRoomId(roomId)
+
+		try {
+			await this.acquireLock(sanitizedRoomId)
+
+			const backupData = this.prepareBackupData(sanitizedRoomId, data)
+			await this.writeBackupFile(sanitizedRoomId, backupData)
+			await this.cleanupOldBackups(sanitizedRoomId)
+
+			return backupData.id
+		} finally {
+			await this.releaseLock(sanitizedRoomId)
+		}
+	}
+
+	/**
+	 * Prepares backup data structure
+	 * @param {number} roomId - The room ID
+	 * @param {object} data - The data to backup
+	 * @return {BackupData} Prepared backup data
+	 */
+	prepareBackupData(roomId, data) {
+		return {
+			id: crypto.randomUUID(),
+			timestamp: Date.now(),
+			roomId,
+			checksum: this.calculateChecksum(data),
+			data,
+			savedAt: data.savedAt || Date.now(),
+		}
+	}
+
+	/**
+	 * Writes backup data to file
+	 * @param {number} roomId - The room ID
+	 * @param {BackupData} backupData - The data to write
+	 */
+	async writeBackupFile(roomId, backupData) {
+		const backupFile = path.join(
+			this.backupDir,
+			`${roomId}_${backupData.timestamp}.bak`,
+		)
+		const tempFile = `${backupFile}.tmp`
+
+		const compressed = await gzip(JSON.stringify(backupData))
+		await fs.writeFile(tempFile, compressed)
+		await fs.rename(tempFile, backupFile)
```

Comment:
What's the advantage of writing the tempFile to disk if we're instantly renaming it?

Reply:
Good points, this was what I searched at that time, not really tested through too :)


**Temporary File Usage**

The temporary file pattern (writing to .tmp then renaming) is actually a common safety practice for file operations, even if the rename happens immediately after. Here's why:
- **Atomic Operations**: The rename operation is atomic on most filesystems, while writeFile is not. This means that if the process crashes during a write, you won't end up with a partially written backup file.
- **Preventing Corruption**: If another process is reading the backup file, the temporary file approach ensures they always get a complete file.
- **Filesystem Caching**: The actual disk write might be buffered by the OS, so the rename provides a guarantee that the write is complete.